Framework/Beamline/inc/MantidBeamline/ComponentInfo.h

```diff
@@ -0,0 +1,48 @@
+#ifndef MANTID_BEAMLINE_COMPONENTINFO_H_
+#define MANTID_BEAMLINE_COMPONENTINFO_H_
+
+#include "MantidBeamline/DllConfig.h"
+#include <vector>
+
+namespace Mantid {
+namespace Beamline {
+
+/** ComponentInfo : Provides a component centric view on the instrument. Indexes
+  are per component.
+
+  Copyright &copy; 2017 ISIS Rutherford Appleton Laboratory, NScD Oak Ridge
+  National Laboratory & European Spallation Source
+
+  This file is part of Mantid.
+
+  Mantid is free software; you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation; either version 3 of the License, or
+  (at your option) any later version.
+
+  Mantid is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+  File change history is stored at: <https://github.com/mantidproject/mantid>
+  Code Documentation is available at: <http://doxygen.mantidproject.org>
+*/
+class MANTID_BEAMLINE_DLL ComponentInfo {
+
+private:
+  std::vector<std::vector<size_t>> m_detectorIndexes;
```

Comment:
@OwenArnold 

Simple version:

```cpp
std::vector<size_t> m_detectorIndices;
std::vector<std::pair<size_t, size_t>> m_ranges;

span detectorIndices(const size_t index) const {
  return span(m_detectorIndices, m_ranges(index));
}
```

`m_ranges` would have size equal to the number of components. We can make that much much shorter by not storing ranges for detectors, because those would be empty. That can be accomplished by defining component indices such that all detectors come first (we discussed that a while ago, for other reasons such as having only one vector of positions in `ComponentInfo`, but not in `DetectorInfo`). The index to access `m_ranges` is then obtained by a simple offset:

```cpp
span detectorIndices(const size_t index) const {
  if(index < m_numberOfDetectors)
    return span();
  return span(m_detectorIndices, m_ranges(index - m_numberOfDetectors));
}
```